vnext/Shared/InspectorPackagerConnection.cpp

```diff
@@ -0,0 +1,176 @@
+      const std::vector<facebook::react::InspectorPage> &pages) {
+    folly::dynamic payload = folly::dynamic::array;
+    for (facebook::react::InspectorPage page : pages) {
+      folly::dynamic pageDyn = folly::dynamic::object;
+      pageDyn["id"] = page.id;
+      pageDyn["title"] = page.title;
+      pageDyn["vm"] = page.vm;
+
+      pageDyn["isLastBundleDownloadSuccess"] = true;
+      pageDyn["bundleUpdateTimestamp"] = "0";
+
+      payload.push_back(pageDyn);
+    }
+    return payload;
+  }
+
+  static folly::dynamic getVMResponseForPackagerPayload(int64_t pageId, std::string &&messageFromVM) {
+    folly::dynamic payload = folly::dynamic::object;
+    payload[InspectorProtocol::InspectorMessage_eventName_wrappedEvent] = messageFromVM;
+    payload[InspectorProtocol::InspectorMessage_PAGEID] = pageId;
+    return payload;
+  }
+
+  static folly::dynamic getVMResponseOnDisconnectForPackagerPayload(int64_t pageId) {
+    folly::dynamic payload = folly::dynamic::object;
+    payload[InspectorProtocol::InspectorMessage_PAGEID] = pageId;
+    return payload;
+  }
+};
+
+} // namespace
+
+RemoteConnection::RemoteConnection(int64_t pageId, const InspectorPackagerConnection &packagerConnection)
+    : m_packagerConnection(packagerConnection), m_pageId(pageId) {}
+
+void RemoteConnection::onMessage(std::string message) {
+  folly::dynamic response = InspectorProtocol::getMessageForPackager(
+      InspectorProtocol::EventType::WrappedEvent,
+      InspectorProtocol::getVMResponseForPackagerPayload(m_pageId, std::move(message)));
+  std::string responsestr = folly::toJson(response);
+  m_packagerConnection.sendMessageToPackager(std::move(responsestr));
+}
+
+void RemoteConnection::onDisconnect() {
+  folly::dynamic response = InspectorProtocol::getMessageForPackager(
+      InspectorProtocol::EventType::Disconnect,
+      InspectorProtocol::getVMResponseOnDisconnectForPackagerPayload(m_pageId));
+
+  std::string responsestr = folly::toJson(response);
+  m_packagerConnection.sendMessageToPackager(std::move(responsestr));
+}
+
+void InspectorPackagerConnection::sendMessageToPackager(std::string &&message) const {
+  m_packagerWebSocketConnection->Send(std::move(message));
+}
+
+void InspectorPackagerConnection::sendMessageToVM(int64_t pageId, std::string &&message) {
+  m_localConnections[pageId]->sendMessage(std::move(message));
+}
+
+InspectorPackagerConnection::InspectorPackagerConnection(std::string &&url) {
+  std::vector<winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult> certExceptions;
+  m_packagerWebSocketConnection =
+      std::make_shared<Microsoft::React::WinRTWebSocketResource>(std::move(url), std::move(certExceptions));
+
+  m_packagerWebSocketConnection->SetOnError([this](const Microsoft::React::IWebSocketResource::Error &err) {
+    assert(false && "Websocket connection failed !");
+  });
+
+  m_packagerWebSocketConnection->SetOnMessage([this](size_t length, const std::string &message, bool isBinary) {
```

Comment:
Capturing by raw `this` seems like it may be unsafe here. Can we use something with weak ownership instead?

Reply:
Thanks! As we do asynchronous read/write raw this can be risky .. Moving the object to be owned and moved
 by shared pointer ..